staging/src/k8s.io/client-go/tools/cache/controller.go

```diff
@@ -359,10 +361,33 @@ func newInformer(
 				switch d.Type {
 				case Sync, Added, Updated:
 					if old, exists, err := clientState.Get(d.Object); err == nil && exists {
-						if err := clientState.Update(d.Object); err != nil {
-							return err
+						oldMeta, err := meta.Accessor(old)
+						if err != nil {
+							return fmt.Errorf("old object has no meta: %v", err)
+						}
+						newMeta, err := meta.Accessor(d.Object)
+						if err != nil {
+							return fmt.Errorf("new object has no meta: %v", err)
+						}
+
+						if oldMeta.GetUID() == newMeta.GetUID() {
```

Comment:
@liggitt this is what I meant.

The deltaFIFO construction and using MetaNamespaceKeyFunc already requires the object to be "meta" (https://github.com/kubernetes/kubernetes/pull/82572/commits/6f02c9e089bc73881b5a956393adc2e0324efb37#diff-1e8e48baa35b3eefb0de35aaa9ba8aa0R349) so I though we can just take UID from it here as well

Reply:
cc @smarterclayton on teaching informers about uids

Reply:
This has such wide reaching implications it would need a KEP or similar level of detail.  Changing controllers at a fundamental level needs to be very carefully thought through.

Reply:
ok, I've removed the generic fix and limited it to RS